### PR TITLE
KEYCLOAK-15244 Require user to agree to 'terms and conditions' during registration

### LINKED ISSUE #8750: Require user to agree to 'terms and conditions' during registration
### Description

When a new user registers, personal data is already collected, stored and processed before the necessary consent is obtained from the user. This violates European data protection regulations and disqualifies the use of Keycloak within the scope of European law. I am astonished that so less requests on this subject have arrived. The only report I found was KEYCLOAK-8447.

Is there any workaround for this?

### Discussion

See https://issues.redhat.com/browse/KEYCLOAK-15244

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fix #8750 

This change adds a `registration-terms-and-conditions` authenticator.
It is disabled by default.
The authenticator allows to force terms and conditions acceptation during user registration.

Obviously, a realm allowing user registration is a prerequisite.

The authenticator must be first enabled in the required actions panel:

Then its requirement must be switched from `DISABLED` to `REQUIRED` in the registration flow:

All of this adds a checkbox that must be checked to complete the registration process:

Screenshot from 2021-04-15 21-06-03

### CODE REVIEW

**Comment by thomasdarimont** [comment]
Wouldn't it have been easier to make `RegistrationUserCreation` configurable and add a "requireTerms" config property?

**Comment by reda-alaoui** [comment]
I followed `Recaptcha` example. The advantage of the separate `FormAction` is greater modularity. Also it may allow to re-use the action in a context different from user registration in the future.

But I would not mind turning this into a `RegistrationUserCreation` configuration option if needed.

**Comment by agemuend** [comment]
We are very interested in this, is it still being worked on?

**Review by stianst — DISMISSED** [review]
Requiring users to accept t&c during registration would be good to have, but please provide a summary of the proposed changes here, and how it should work.

**Comment by pdomineaux** [comment]
Hello, thanks for the amazing work, any updates on that GDPR compliant feature?

**Comment by pachisi456** [comment]
Any updates on this? Would be quite interested in using this soon.

**Comment by reda-alaoui** [comment]
Ping :pray:

**Comment by tbdaniel27** [comment]
:+1: 

This is actually a legal requirement in Germany. Forcing the user to accept the TOS after registration via the required action is too late because the contract was already created. Therefore, please consider accepting this feature :pray:

**Comment by wojtekkul** [comment]
Any updates?

**Comment by RANSoftRA** [comment]
Would be nice to have this in the next version.

**Comment by reda-alaoui** [comment]
Ping :cry:

**Comment by adamalexandru4** [comment]
We want this feature!

**Comment by thuemmler-zeppelin** [comment]
@reda-alaoui I think the feature would be able to be merged if you fulfilled @stianst changes request. It is just to provide a summary and how it works. 
I think this would be a powerful highly needed feature for european needs

**Comment by reda-alaoui** [comment]
> @reda-alaoui I think the feature would be able to be merged if you fulfilled @stianst changes request. It is just to provide a summary and how it works. I think this would be a powerful highly needed feature for european needs

@thuemmler-zeppelin , the content of https://github.com/keycloak/keycloak/pull/7938#issue-859160934 has already been updated in `January 30 2022`  following https://github.com/keycloak/keycloak/pull/7938#pullrequestreview-805712249 .

**Comment by thuemmler-zeppelin** [comment]
> > @reda-alaoui I think the feature would be able to be merged if you fulfilled @stianst changes request. It is just to provide a summary and how it works. I think this would be a powerful highly needed feature for european needs
> 
> @thuemmler-zeppelin , the content of [#7938 (comment)](https://github.com/keycloak/keycloak/pull/7938#issue-859160934) has already been updated on `January 30, 2022` following [#7938 (review)](https://github.com/keycloak/keycloak/pull/7938#pullrequestreview-805712249) posted on `November 15, 2021`.
> 
> From my POV, the change request has been fulfilled a long time ago.

@reda-alaoui  Oh sorry i did not see that one. Maybe someone (@stianst ) can resolve this as resolved then? ;-)

**Comment by mabartos** [comment]
@mposolda @stianst @rmartinc Guys, do you think you have some time, this or next sprint, to review this PR? :))

**Comment by reda-alaoui** [comment]
:sleeping:

**Review by rmartinc — CHANGES_REQUESTED** [review]
Thanks @reda-alaoui for the PR and really sorry for the delay!

It LGTM  and I think it's a very good idea. Current terms and conditions need the user created as they are implemented as a user action. But for sure you have to change the import to `jakarta` or it won't compile in main now (it's our fault for not reviewing it sooner).

Please rebase the branch to put PR changes on top of current main and modify the import. Also squash the commits into a singe one once the PR is being reviewed.

I have a doubt about if we usually modify the `registration` flow in upgrades (for existing realms). I have checked a bit and I think the answer is no. The old flows remain as it were before. Nevertheless I will ask to the team and I'll confirm you the next week.

Have a nice weekend!

File: themes/src/main/resources/theme/base/login/register-user-profile.ftl [hunk 1]
```diff
@@ -49,6 +49,32 @@
...
+					</div>
+					<#if messagesPerField.existsError('termsAccepted')>
+						<div class="${properties.kcLabelWrapperClass!}">
+                            <span id="input-error-terms-accepted" class="${properties.kcInputErrorMessageClass!}" aria-live="polite">
+                                ${kcSanitize(messagesPerField.get('termsAccepted'))?no_esc}
+                            </span>
+						</div>
+					</#if>
+				</div>
+			</#if>
```
[the remark this answers is not in the document]
Comment (reda-alaoui) [reply themes/src/main/resources/theme/base/login/register-user-profile.ftl:77]: It looks like to me that the indentation is already at the same level than the recaptcha  block. And what do you mean by OCD?

[the remark this answers is not in the document]
Comment (rmartinc) [reply themes/src/main/resources/theme/base/login/register-user-profile.ftl:77]: :smile:  My [ODC](https://en.wikipedia.org/wiki/Obsessive%E2%80%93compulsive_disorder), I think the new part is to the right, isn't it?

```xml
			</#if>
            
            <#if recaptchaRequired??>
```

File: themes/src/main/resources/theme/base/login/register-user-profile.ftl [hunk 2]
```diff
@@ -49,6 +49,32 @@
 		            </#if>
                 </#if>  
             </@userProfileCommons.userProfileFormFields>
+
+            <#if termsAcceptanceRequired??>
```
Comment (pedroigor) [inline themes/src/main/resources/theme/base/login/register-user-profile.ftl:53]: Would make sense to create a macro and import/re-use the template on both registration pages?

**Comment by rmartinc** [comment]
@reda-alaoui I asked and some other colleagues also think that upgrade modifications are not needed (the registration flow will remain the same in the existing realms). So nothing more needed for that part.

Just one more idea from me. Maybe it's recommendable to add some documentation about this. I think a good point would be after [Enabling user regitration](https://www.keycloak.org/docs/latest/server_admin/#proc-enabling-user-registration_server_administration_guide) in the [admin guide](https://github.com/keycloak/keycloak/tree/main/docs/documentation/server_admin/). Adding one more topic like `Require user to agree to 'terms and conditions' during registration`. And you just comment the same you did here in the issue. WDYT?

**Comment by rmartinc** [comment]
@reda-alaoui The code is OK to me now, but what do you think about the documentation? Do you want to add the new topic for the doc? We can do it in a follow-up issue if you don't want to be involved with doc changes...

**Comment by rmartinc** [comment]
@reda-alaoui Thanks! Take all the time you need, I was not rushing you to do it.

**Comment by reda-alaoui** [comment]
@rmartinc , documentation added :+1:

**Review by pedroigor — DISMISSED** [review]
LGTM

**Review by rmartinc — DISMISSED** [review]
Thanks @reda-alaoui ! LGTM too! I'm adding @andymunro for the documentation changes.

File: docs/documentation/server_admin/topics/users/proc-requiring-tac-agreement-at-registration.adoc [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+//
+// con-user-registration.adoc
+
+[id="proc-requiring-tac-agreement-at-registration_{context}"]
+= Requiring user to agree to terms and conditions during registration
+
+[role="_abstract"]
+You can condition a user registration to his agreement to your terms and conditions.
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-requiring-tac-agreement-at-registration.adoc:9]:
```suggestion
For a user to register, you can require agreement to your terms and conditions.
```

File: docs/documentation/server_admin/topics/users/proc-requiring-tac-agreement-at-registration.adoc [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+.Registration form with required terms and conditions agreement
+image:images/registration-form-with-required-tac.png[]
+
+.Prerequisite
+* User registration is enabled.
+* Terms and conditions required action is enabled
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-requiring-tac-agreement-at-registration.adoc:16]:
```suggestion
* Terms and conditions required action is enabled.
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good, @pedroigor.  Just one change to suggest.

**Review by rmartinc — APPROVED** [review]
@pedroigor This seems to be ready! Thanks a lot @reda-alaoui!

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@reda-alaoui Thanks. nice addition.

**Comment by danieliu456** [comment]
@reda-alaoui 

Is there any possibility to set user attributes when the user accepts terms and conditions with this new flow? 
Or did I misunderstand something? Because I can't see any attributes on the user after testing this functionality?

P.S. @reda-alaoui Thank you for this awesome feature and contribution! <3

Old flow: 
```java
    @Override
    public void processAction(RequiredActionContext context) {
        // Keycloak 21.0.0 changed the user attribute name from lowercase to uppercase
        // this change was reverted, but it is still possible some attributes created
        // in Keycloak 21.0.0 will be present in the database, we need to remove it too.
        // See https://github.com/keycloak/keycloak/issues/17277 for more details
        context.getUser().removeAttribute(USER_ATTRIBUTE.toUpperCase());

        if (context.getHttpRequest().getDecodedFormParameters().containsKey("cancel")) {
            context.getUser().removeAttribute(USER_ATTRIBUTE);
            context.failure();
            return;
        }

        context.getUser().setAttribute(USER_ATTRIBUTE, Arrays.asList(Integer.toString(Time.currentTime())));

        context.success();
    }
```

New:
```java
	@Override
	public void success(FormContext context) {

	}
```

**Pull request merged**